### PR TITLE
 WritePrepared Txn: Duplicate Keys, Txn Part

### PR DESCRIPTION
**Pull request by maysamyabandeh**
This patch takes advantage of memtable being able to detect duplicate <key,seq> and returning TryAgain to handle duplicate keys in WritePrepared Txns. Through WriteBatchWithIndex's index it detects existence of at least a duplicate key in the write batch. If duplicate key was reported, it then pays the cost of counting the number of sub-patches by iterating over the write batch and pass it to DBImpl::Write. DB will make use of the provided batch_count to assign proper sequence numbers before sending them to the WAL. When later inserting the batch to the memtable, it increases the seq each time memtbale reports a duplicate (a sub-patch in our counting) and tries again.

### CODE REVIEW

File: utilities/transactions/write_prepared_txn.cc [hunk 1]
```diff
@@ -63,6 +62,56 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
...
+    if (lhs.first != rhs.first) {
+      return lhs.first < rhs.first;
+    }
+    return lhs.second.compare(rhs.second) < 0;
+  }
+};
+// Count the number of sub-batches inside a batch. A sub-batch does not have
+// duplicate keys.
+struct SubBatchCounter : public WriteBatch::Handler {
+  std::set<std::pair<uint32_t, Slice>, CFKeyComparator> keys_;
```
Comment (yiwu-arbug) [inline utilities/transactions/write_prepared_txn.cc]: would a hash set be more efficient? There's a `SliceHasher` class you can reuse.

File: utilities/transactions/write_prepared_txn.cc [hunk 2]
```diff
@@ -63,6 +62,56 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
...
+  size_t batches_;
+  SubBatchCounter() : batches_(1) {}
+  size_t BatchCnt() { return batches_; }
```
Comment (yiwu-arbug) [inline utilities/transactions/write_prepared_txn.cc]: nit: s/BatchCnt/BatchCount/ ? No need to shorten names that's not long..

File: utilities/transactions/write_prepared_txn.cc [hunk 3]
```diff
@@ -63,6 +62,56 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
...
+  void AddKey(uint32_t cf, const Slice& key) {
+    auto size = keys_.size();
+    keys_.insert({cf, key});
```
Comment (yiwu-arbug) [inline utilities/transactions/write_prepared_txn.cc]: you can check the return of `insert` to see if it is duplicated.

File: utilities/transactions/write_prepared_txn_db.h [hunk 4]
```diff
@@ -401,30 +401,56 @@ class WritePreparedCommitEntryPreReleaseCallback : public PreReleaseCallback {
...
+  }
 
   virtual Status Callback(SequenceNumber commit_seq) {
     assert(includes_data_ || prep_seq_ != kMaxSequenceNumber);
+    const uint64_t last_commit_seq = LIKELY(data_batch_cnt_ <= 1)
+                                         ? commit_seq
+                                         : commit_seq + data_batch_cnt_ - 1;
     if (prep_seq_ != kMaxSequenceNumber) {
-      db_->AddCommitted(prep_seq_, commit_seq);
+      if (LIKELY(prep_batch_cnt_ == 1)) {
```
Comment (yiwu-arbug) [inline utilities/transactions/write_prepared_txn_db.h]: why you need the if-else block? Compiler will likely to expand the loop for you (at least with -O3).

File: db/write_batch.cc [hunk 5]
```diff
@@ -1301,18 +1345,32 @@ class MemTableInserter : public WriteBatch::Handler {
         perform_merge = false;
       } else {
         // 3) Add value to memtable
-        mem->Add(sequence_, kTypeValue, key, new_value);
+        bool mem_res = true;
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply db/write_batch.cc]: yes. the new approach took more time to implement but entirely and cleanly takes care of any duplicate key problem.

File: utilities/transactions/write_prepared_txn.cc [hunk 6]
```diff
@@ -63,6 +62,56 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
   return write_batch_.NewIteratorWithBase(column_family, db_iter);
 }
 
+struct CFKeyComparator {
+  bool operator()(const std::pair<uint32_t, Slice>& lhs,
+                  const std::pair<uint32_t, Slice>& rhs) const {
+    if (lhs.first != rhs.first) {
+      return lhs.first < rhs.first;
+    }
+    return lhs.second.compare(rhs.second) < 0;
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply utilities/transactions/write_prepared_txn.cc]: Good catch. I will update the PR.
Yeah the user comparator should not return not-equal for two identical binary but this comparator should return equal exactly when the user comparator in memtable does so.

File: utilities/transactions/write_prepared_txn.cc (hunk 1 above)
Comment (maysamyabandeh) [reply utilities/transactions/write_prepared_txn.cc]: yeah makes sense.

File: utilities/transactions/write_prepared_txn.cc (hunk 2 above)
Comment (maysamyabandeh) [reply utilities/transactions/write_prepared_txn.cc]: sure.

File: utilities/transactions/write_prepared_txn.cc (hunk 3 above)
Comment (maysamyabandeh) [reply utilities/transactions/write_prepared_txn.cc]: sure

File: utilities/transactions/write_prepared_txn.cc [hunk 7]
```diff
@@ -63,6 +62,76 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
...
+  }
+};
+struct SetComparator {
+  SetComparator() : user_comparator_(BytewiseComparator()) {}
+  SetComparator(const Comparator* user_comparator)
```
Comment (yiwu-arbug) [inline utilities/transactions/write_prepared_txn.cc]: `explicit`

File: utilities/transactions/write_prepared_txn.cc [hunk 8]
```diff
@@ -63,6 +62,76 @@ Iterator* WritePreparedTxn::GetIterator(const ReadOptions& options,
...
+
+ private:
+  const Comparator* user_comparator_;
+};
+// Count the number of sub-batches inside a batch. A sub-batch does not have
+// duplicate keys.
+struct SubBatchCounter : public WriteBatch::Handler {
+  SubBatchCounter(std::map<uint32_t, const Comparator*>& comparators)
+      : comparators_(comparators), batches_(1) {}
+  std::map<uint32_t, const Comparator*>& comparators_;
```
Comment (yiwu-arbug) [inline utilities/transactions/write_prepared_txn.cc]: you can still use unordered_map.

Comment (maysamyabandeh) [reply utilities/transactions/write_prepared_txn.cc]: given that there will be a few cf in each txn (perhaps only 1), and given that cf is integer i think binary search tree would be as good as hash map if not faster.

Comment (yiwu-arbug) [reply utilities/transactions/write_prepared_txn.cc]: ok. yeah it is really hard to tell which one is better when the size is really small.

**Pull request closed without merging**